Added warnings to command module

willthames opened the pull request:
Generate warnings when users are shelling out to commands
rather than using modules

Can be turned off on a per-action line with the documented
warn=False flag.

Print out warnings using the standard playbook callbacks.

willthames committed:
Added warnings to command module

Generate warnings when users are shelling out to commands
rather than using modules

Can be turned off on a per-action line with the documented
warn=False flag. Can be turned off globally using
command_warnings = False in ansible config file.

Print out warnings using the standard playbook callbacks.

Created some additional tests in TestRunner.test_command
and also a demonstration playbook.

willthames commented:
I'm not necessarily expecting this to get accepted on first pass - consider it an RFC. 

All existing tests pass, can probably add some new ones if doing this is desirable.

mpdehaan commented:
I like this idea.

I'm thinking maybe also would want a configuration setting in ansible.cfg that can ignore these globally?

willthames commented:
Yep - should be easy enough to implement. One for tomorrow though - nearly midnight here!

jctanner commented:
@willthames Sorry for the delay on reviewing this, but it appears to be breaking one of the runner tests ...

```
======================================================================
FAIL: test_command (TestRunner.TestRunner)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jtanner/ansible/test/TestRunner.py", line 154, in test_command
    assert len(result.get('warnings')) == 0
AssertionError: 
    {u'changed': True, u'end': u'2014-01-28 10:30:31.511526', u'stdout': u'/home/jtanner', u'cmd': u'/bin/echo $HOME ', u'rc': 0, u'start': u'2014-01-28 10:30:31.509586', u'stderr': u'', u'delta': u'0:00:00.001940', 'invocation': {'module_name': 'shell', 'module_args': u'/bin/echo $HOME'}, u'warnings': [u'Consider using template or lineinfile module rather than running echo']} = self._run('shell', ["/bin/echo", "$HOME"])
    assert 'failed' not in {u'changed': True, u'end': u'2014-01-28 10:30:31.511526', u'stdout': u'/home/jtanner', u'cmd': u'/bin/echo $HOME ', u'rc': 0, u'start': u'2014-01-28 10:30:31.509586', u'stderr': u'', u'delta': u'0:00:00.001940', 'invocation': {'module_name': 'shell', 'module_args': u'/bin/echo $HOME'}, u'warnings': [u'Consider using template or lineinfile module rather than running echo']}
    assert {u'changed': True, u'end': u'2014-01-28 10:30:31.511526', u'stdout': u'/home/jtanner', u'cmd': u'/bin/echo $HOME ', u'rc': 0, u'start': u'2014-01-28 10:30:31.509586', u'stderr': u'', u'delta': u'0:00:00.001940', 'invocation': {'module_name': 'shell', 'module_args': u'/bin/echo $HOME'}, u'warnings': [u'Consider using template or lineinfile module rather than running echo']}['rc'] == 0
>>  assert len({u'changed': True, u'end': u'2014-01-28 10:30:31.511526', u'stdout': u'/home/jtanner', u'cmd': u'/bin/echo $HOME ', u'rc': 0, u'start': u'2014-01-28 10:30:31.509586', u'stderr': u'', u'delta': u'0:00:00.001940', 'invocation': {'module_name': 'shell', 'module_args': u'/bin/echo $HOME'}, u'warnings': [u'Consider using template or lineinfile module rather than running echo']}.get('warnings')) == 0
```

willthames commented:
There was a failing test and fixing the failing test made me notice a minor logic error (warnings should be an empty list, rather than None, when there are no warnings). 

This is now fixed, all tests pass again (not sure how the failure got introduced, as I worked on 40a44ce6 while first looking at the tests for this change so I wasn't ignoring them!) .

jctanner commented:
@willthames I attempted to fix the current merge conflicts with these, but one of the tests appears to be broken ...

```
======================================================================
ERROR: test_command (TestRunner.TestRunner)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jtanner/ansible/test/TestRunner.py", line 174, in test_command
    assert len(result.get('warnings')) == 1
TypeError: object of type 'NoneType' has no len()

----------------------------------------------------------------------
Ran 96 tests in 22.525s

FAILED (SKIP=1, errors=1)
make: *** [tests] Error 1
```

If you can take a look, fix the conflict and the test(s), I'll merge this in.

willthames commented:
That looks like the problem I fixed in the push before last.

I just rebased against the latest devel branch, fixing the merge conflict relating to adding NO_LOG and warn as arguments to command, and no test failures for me.

I've repushed the result, so please have another go.

jimi-c commented:
In queue for review, we're going to have to at least move the tests into the new integration test suite, as the old nose tests are no longer used.

willthames commented:
@jimi-c I've added some tests into the new integration test suite and repushed.

mpdehaan commented:
Obviously there has been some delay in this one. It does seem that this may produce some false positives that users may find rough if they are doing things like running yum commands that the module does not surface.

As such, I think the warn=yes/no flag is useful, but also it would be ideal if warn=yes was auto-added to the command module based on a default in settings.py if warn=yes/no is not already supplied.

The easiest way to control this is probably to move this logic up into the "normal" module, which is something I'm going to look into.

mpdehaan commented:
By which I mean, we should be able to repurpose most of the above code, just moving it up into normal.py -- I'm looking at this now.

mpdehaan commented:
Sorry, reading comprehension fail, this does this.  Awesome.   Looking now.

mpdehaan commented:
I've reduced just a few of the warnings: shell cp is useful for local file copy, and echo I use occasionally for very silly test examples. 

Since I wanted those to work with the default, I've removed those, but am up for a discussion.

I really think we could make this configurable by moving it to a plugin structure or perhaps ansible.cfg or another file, with defaults, which I'm open to.

Good discussion for ansible-devel.

In all, longer term, I _DO_ want more checking and awareness in ansible like this, and we also need to improve some various classes to do a bit more checking about arguments that are used together, whereas some fields in playbooks are still treated a bit like datastructures.  We'll get there over time.

I'm leaving this on by default for now, but we _might_ need to turn it off by default, TBD.  We'll find out :)

mpdehaan commented:
Merged, thanks!

mpdehaan commented:
Posted question to list about whether we think this should be on by default or off BTW.  Thoughts welcome.
